docs/proposals/202105256-cluster-class-and-managed-topologies.md

```diff
@@ -229,8 +262,182 @@ type LocalObjectTemplate struct {
+type VariableDefinitionSchemaClass struct{
+  // OpenAPIV3Schema defines the schema of a variable via OpenAPI v3
+  // schema. The schema is a subset of the schema used in
+  // Kubernetes CRDs.
+  OpenAPIV3Schema JSONSchemaProps `json:"openAPIV3Schema"`
+}
+```
+
+The schema implementation will be built on top of the [Open API schema embedded in Kubernetes CRDs](https://github.com/kubernetes/apiextensions-apiserver/blob/master/pkg/apis/apiextensions/types_jsonschema.go).
+To keep the implementation as easy and user-friendly as possible we will only implement the following feature set 
+for now (until further use cases emerge):
+- Basic types:
+    - boolean, integer, number, string
+    - nullable, i.e. each of the basic types can be set to null
+- Basic validation, e.g. format, minimum, maximum, pattern, required, ...
+- Defaulting
+    - Defaulting will be implemented based on the CRD structural schema library and thus will have the same feature set 
+      as CRD defaulting. I.e., it will only be possible to use constant values as defaults.
+    - Through clusterctl we have an additional layer of templating for the Cluster resource which allows defaulting of 
+      Cluster variables via environment variables.
+
+**PatchClass**
+
+```golang
+// PatchClass defines a patch which is applied to customize the referenced templates.
+type PatchClass struct {
+  // Name of the patch.
+  Name string `json:"name"`
+ 
+  // Definitions define the patches inline.
+  // Note: Patches will be applied in the order of the array.
+  Definitions []PatchDefinition `json:"definitions,omitempty"`
+}
+ 
+// PatchDefinition defines a patch which is applied to customize the referenced templates.
+type PatchDefinition struct {
+  // Selector defines on which templates the patch should be applied.
+  Selector PatchDefinitionSelector `json:"selector"`
+ 
+  // JSONPatches defines the patches which should be applied on the templates
+  // matching the selector.
+  // Note: Patches will be applied in the order of the array.
+  JSONPatches []JSONPatchDefinition `json:"jsonPatches"`
+}
+```
+
+**Note**: We are considering to add a field to allow defining optional patches. Optional patches could be 
+activated by the user on a per-cluster basis.
+
+**PatchDefinitionSelector**
+
+```golang
+// PatchDefinitionSelector defines on which templates the patch should be applied.
+// Note: Matching on APIVersion and Kind is mandatory, to enforce that the patches are 
+// written for the correct version. The version of the references may be automatically
+// updated during reconciliation if there is a newer version for the same contract.
+type PatchDefinitionSelector struct {
+  // APIVersion filters templates by apiVersion.
+  APIVersion string `json:"apiVersion"`
+ 
+  // Kind filters templates by kind.
+  Kind string `json:"kind"`
+ 
+  // MatchResources selects templates based on where they are referenced.
+  MatchResources PatchMatchResources `json:"matchResources"`
```

Comment:
I wonder if something like `matchContext` would be better.

Reply:
Resources sounds good to me as-is, given that it's a known Kubernetes concept

Reply:
Was just wondering because we're kind of matching **on** the context where those resources are used and not the resources themselves